### ISSUE TITLE
Is this plugin still working for you?

### ISSUE THREAD

**Issue by galitus** [MattRCole/plugin.video.piped#1]
Hey,

I tried to use your plugin, but it didn't work from the start
because the URL while sharing from my phone with newpipe or libretube is just different.
There is no /play?video_id= .
It is always /watch?v=..... .
So, I modified it on my end to extract the video id and I got it to proceed 

`DEBUG <general>: StartScript - calling plugin Piped Video Player('plugin://plugin.video.sendtokodi/','14','?https://piped.video/watch?v=Ofgc9vI9nZE','resume:true')`

Yeah, I also had to change the plugin name to “sendtokodi” to act on this forward from the phone.

But then I get this stack trace:

```
ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'urllib.error.HTTPError'>
                                                   Error Contents: HTTP Error 500: Internal Server Error
                                                   Traceback (most recent call last):
                                                     File "~/.kodi/addons/plugin.video.sendtokodi/resources/lib/main.py", line 194, in <module>
                                                       router(sys.argv[2][1:], action)
                                                     File "~/.kodi/addons/plugin.video.sendtokodi/resources/lib/main.py", line 185, in router
                                                       play_video(params['v'])
                                                     File "~/.kodi/addons/plugin.video.sendtokodi/resources/lib/main.py", line 126, in play_video
                                                       get_playlist_ready(video_playlist, base_url=base_url, base_dir=base_dir, hls_path="video-index.m3u8")
                                                     File "~/.kodi/addons/plugin.video.sendtokodi/resources/lib/main.py", line 54, in get_playlist_ready
                                                       playlist = m3u8.load(resolved_url)
                                                     File "~/.kodi/addons/script.module.m3u8/lib/m3u8/__init__.py", line 48, in load
                                                       content, base_uri = http_client.download(uri, timeout, headers, verify_ssl)
                                                     File "~/.kodi/addons/script.module.m3u8/lib/m3u8/httpclient.py", line 17, in download
                                                       resource = opener.open(uri, timeout=timeout)
                                                     File "/usr/lib/python3.10/urllib/request.py", line 525, in open
                                                       response = meth(req, response)
                                                     File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
                                                       response = self.parent.error(
                                                     File "/usr/lib/python3.10/urllib/request.py", line 563, in error
                                                       return self._call_chain(*args)
                                                     File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
                                                       result = func(*args)
                                                     File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
                                                       raise HTTPError(req.full_url, code, msg, hdrs, fp)
                                                   urllib.error.HTTPError: HTTP Error 500: Internal Server Error
                                                   -->End of Python script error report<--

```

Any thoughts? What went wrong? 
Maybe you experienced that problem while creating it in the first place?

Thanks in advance

**Comment by MattRCole** [comment]
Hey! Glad to see that someone is still trying to use this plugin!

Its been a while since I've tried to use the plugin myself because it started breaking for some reason, but I think I have some context as to why this is happening:

To start, `piped` only provides direct video+audio URLs if the quality of the video is 720p or below, for anything higher they provide a few options:

1. A link to an audio stream, as well as several links to different video streams of varying quality.
2. HLS Manifests (`m3u8` playlists) (HLS: HTTP Live Streaming)
3. DASH similar to HLS, but, while the option is implied to be provided by `piped`, at the time of writing the plugin, I never saw it given, so I never used it.

I decided to use the HLS option because, while more complex, it allows you to seek around the video fairly flawlessly.

`piped` typically provides a link to a parent HLS manifest first that links to several other child HLS manifests (IE: the first manifest just has a list of different quality/bit-rate options as well as audio stream options)

Those "child" manifests then actually point to small segments of video/audio that is stitched together on-the-fly by the video player which allows the end user (you and me) to watch the video and hear the audio from the YouTube video before the whole video is loaded. Neat.

I'm giving all this context for either A) you aren't familiar with HLS streaming and it'll be helpful in explaining the bug, or B) you know more than me, and can point out errors in my reasoning if there are any. Let's move onto the janky hack I needed to do to get HLS to work

## My janky hack

Ffmpeg is the player I used and it supports HLS manifests, which is great. What is not so great is that ffmpeg seems to have a bug: if it gets a link to an HLS manifest that references a child HLS manifest, it can take up to 3-4 minutes for ffmpeg to load the child HLS manifest file. I'm not sure why that is, these files are vary small, and the bug was present when I built ffmpeg from source, so it wasn't a matter of Kodi using an outdated version of ffmpeg.

Since `piped` provides a parent HLS manifest that references child HLS manifests, this bug was an issue.

My solution to this was to go through every HLS manifest in code withe the `m3u8` plugin, download all child manifests manually, and then change the URLs of the parent manifest to just be file paths to the downloaded child manifest.

## the bug

Now we finally get to this bug.

The error is being thrown when the script tries to download a child HLS manifest. Since piped is returning a `500` http response code, there's a small chance that its a bug on `piped`'s end, but the more likely culprit is something I've done in the code (maybe the URL I make is malformed?).

I'm not a `kodi` add-on developer, I made this for personal use and decided to upload it here in the hopes that I could get someone who really knows what they're doing to either help out with development or clone what I've done and make a better plugin. 

When the plugin broke for me, I found it hard to track down why (I use LibreElec as my kodi OS, so its hard to find the plugin logs), so I eventually set up a much more janky solution for watching YouTube videos and abandoned the plugin.

I am very interested in trying to fix the plugin and might have some free time for once in my life this weekend where I can look into what's happening and try to fix it.

If you want to as well, or want to add a feature to the plugin (like those changes you mentioned) I'm more than open to a pull request.

EDIT: Some clarity and spelling (I was on mobile)

**Comment by galitus** [comment]
Hey,

at first thanks for the fast and length response!

But I have to disappoint you, I'm not a coder. 
I know my way around in some languages to modify 
it to my needs and script something here and there but also not fully qualified
and experienced in this case. : )

My approach to this would be, to check how yt-dlp or pytube are doing it because it is also python 
and maybe the needed part can just be imported and broken of at the point where you/we can feed it to
ffmpeg/inputstream-adaptive to play the video? (but perhaps I'm too optimistic at that part)
If this could work, they will also figure it out to fix something in the future when youtube/piped change something and breaks it.

youtube_extract (here on github) used yt-dlp for extracting metadata from channels and videos, 
potentially this could be helpful to extract not metadata, but the URL needed to play it? : )

I don't even know if it is possible or allowed to use such external libs in kodi?
Like I said, I have no idea about that stuff. 

Here some more information about the error, I printed out the resolved_url it was using and tried to access it in my browser and the response was "Invalid qhash provided". (I don't know what it means or better what's wrong.)
The url included this "&qhash=2006eeef" so, there was one submitted but it seems a wrong one. 
I didn't know where it was coming from.

Btw I use kodi baremetal, debug was on and I edited the file for debug printing and modifying directly in the kodi folder.

The important part I modified, was this:

```
from urllib.parse import urlencode, parse_qsl, urlparse
```
adding **urlparse** and that:
```   
    params1 = urlparse(paramstring)
    params = dict(parse_qsl(params1.query))

    action = 'watch'
    if action:
	if action == 'watch':
            play_video(params['v'])
        else:
            raise ValueError('Invalid paramstring: {}!'.format(paramstring))
```
(I never used git for a pull request and I just wanted to give you all the information I have at the moment.)

Someone also forked your plugin and also did some modifications for using newpipe to send the URL to the plugin
BUT I couldn't even get it installed in the first place.  (kodi isn't helpful there because even in the logfile it isn't mentioned why it failed .....)
The action part which I also modified is perhaps useful. 

https://github.com/jonhermansen/plugin.video.piped/tree/small-improvements

Greetings, and thanks for trying to make a plugin in the first place!

**Comment by MattRCole** [comment]
Thanks for the info!

I will try yt-dlp/pytube if I can't get this working. My hope was that, by using piped.video, all of the hard work of reverse-engineering youtube's front-end/back-end could be avoided.

Thanks for the additional debug info. I have some time now to look into this, so I'll keep you updated on any progress!

**Comment by MattRCole** [comment]
I have a few updates after working at this today:

1) I was able to recreate your error, but only on certain videos. It looks like this is most likely an error with `piped` and not specifically this plugin.
2) I have not been able to get anything to play using the `ffmpegdirect` inputstream. This was extremely annoying, as playing the `index.m3u8` file would usually work (either via `ffplay` or `vlc`), and I saw no errors in the log file.
    - I then tried `inputstream.adaptive`, but without any success or logs either.
3) `pipedapi.kavin.rocks` seems to randomly fail, I switched to `https://pipedapi.syncpundit.io` and had less failures.
4) I figured out a way to trigger the addon via `JSON RPC` using `curl`, which was extremely nice for testing (some editing of the command may be required):

```shell
curl -X POST --header 'Content-Type: application/json' http://localhost:8080/jsonrpc -d '{"jsonrpc": "2.0", "method": "Addons.ExecuteAddon", "params": { "addonid": "plugin.video.piped", "params": "/play?video_id=[VIDEO-ID]", "wait": false }, "id": 1232 }'
```


In conclusion: I was not able to get this plugin working. Without any useful debug information it is going to be very difficult to make any progress.
